File: spiralMatrix_54.cpp
// Time Complexity : O(MN)
// Space Complexity :O(1)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no

class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        if(matrix.empty() || matrix.size() == 0){
            vector<int> x;
            return x;
        }
        int m = matrix.size();
        int n = matrix[0].size();
        int top=0, left=0;
        int bottom = m-1, right = n-1;
        vector<int> arr;
        while(top <= bottom && left <= right){
            //from left to right
            for(int i = left; i <= right; i++){
                arr.push_back(matrix[top][i]);
            }
            top++;

            //top to bottom
            for(int i = top; i <= bottom; i++){
                arr.push_back(matrix[i][right]);
            }
            right--;
            
            //right to left
            if(top <= bottom){
                for(int i = right; i >= left; i--){
                    arr.push_back(matrix[bottom][i]);
                }
                bottom--;
            }
            
            //bottom to top
            if(left <= right){
                for(int i = bottom; i >= top; i--){
                    arr.push_back(matrix[i][left]);
                }
                left++;
            }
        }
        return arr;
    }
};

File: Spiral-matrix_54.java
// Time Complexity : O(m*n) size of matrix
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Three line explanation of solution in plain english
//keep track of top row and bottom row already traversed similarly keep track of left column and right column already traversed
//iterate over the matrix keeping the index between top and bottom row and left and right index

// Your code here along with comments explaining your approach

class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        int tr = 0; //top row tracker
        int br = matrix.length-1; //bottom row tracker
        int lc = 0; //left column tracker
        int rc = matrix[0].length-1; //right column tracker

        List<Integer> res = new ArrayList<Integer>();

        while(tr<=br && lc <= rc) {
            int currCol = lc;
            int currRow = tr;
            while(currCol<=rc) {
                res.add(matrix[tr][currCol++]);
            }
            if (tr == br) break;

            tr++;
            currRow = tr;
            currCol = rc;
            while(currRow<=br) {
                res.add(matrix[currRow++][currCol]);
            }

            rc--;
            currCol = rc;
            currRow = br;
            while(currCol>=lc) {
                res.add(matrix[currRow][currCol--]);
            }

            br--;
            currRow = br;
            currCol = lc;
            while(currRow >=tr && rc >= lc) {
                res.add(matrix[currRow--][currCol]);
            }

            lc++;

        }
        return res;
    }
}
